### DISCUSSION TITLE
Run hosts file monitoring as a Windows service?

### DISCUSSION THREAD

**Discussion by Marakai** [derekhe/msfs2020-google-map#144] [Ideas]
Hello!

I've been excited about this tool since I saw the Youtube videos about it - and by now you've seen a few of my pull requests.

I'm an avid simmer and also write Python code professionally *and* I live in an area where Bing Maps are 10(!) years out of date. Whereas Google Maps are updated every 6 months or so. You can see why I am keenly interested in this.

My goals are to help you with making this app as stable and easy to use for all users as possible. One issue I've noticed a number of users report both here and on Youtube is problems that occur with the changes to the `hosts` file in Windows.

My idea (and I am happy to implement it and send you the PR):

- Change the component that modifies the Windows `hosts` file to a Windows service
- With a Windows service you can configure it so that Windows itself monitors is and restarts it if/when it crashes (but not when the user stops/terminates it).
- Windows has the necessary permissions to edit the `hosts` file
- The service will check every N minutes (N=1 minute or user-defined) whether FS2020 and the proxy are running.
- If they are NOT running, it will check that the redirect entries in `hosts` are *removed* and remove them if they are still in the file
- If FS2020 and the proxy are running, it do the opposite: it will check that the entries exist and add them if they don't
- This will ensure that users only have these entries or not based on whether they are actually running the flight simulator with the app and will always be there to clean up if necessary.

The PRs I've sent you are in preparation to allow refactoring to make the code more modular and allow the required methods to be extracted into the Windows service.

An additional idea I am pursuing is to have the ability to run some of the services (proxy and nginx) running on another machine, or in a container - or even in the cloud!

**Comment by derekhe** [comment]
Thanks for your PR.

I see lots of user have issue with hosts file. Most of the issue is caused by antivirus to prevent writing to hosts files. My thought on this mod is user can run it without install any addition stuff and can disable it easily. The concern using a windows service is that it needs to modify the system to inject the windows service and uninstallation maybe another issue.

The current issue running python script on user system is that it can't do clean up if user close the command prompt window, user had to press "stop" button to restore it. I'm working on a version using electron (see the electron branch) and it will fix this issue. User should turn on the mod and the hosts file should be changed and after they close the app, the hosts file should be restored.

**Comment by Marakai** [comment]
You're absolutely right on the implications of running a Windows service, but many users are very familiar with just using an installer and clicking Next, Next, Finish - and "magic" happens. They may actually be less familiar with going into an unzipped directory and doing various things in there and if something goes wrong they are lost when their system is left in a state they simply don't understand what's wrong.

I'm just looking at it from that point of view (it's what I have to at work :-) 

I think the issue is that if the program terminates unexpectedly then the `hosts` file may be left in an undefined state. Ideally, the exception handling should deal with every such possibility, but we all know that's often impossible.
